### ISSUE TITLE
improve R.wrap examples

### ISSUE THREAD

**Issue by davidchambers** [ramda/ramda#1387]
Current examples:

> ```javascript
> var greet = function(name) {return 'Hello ' + name;};
>
> var shoutedGreet = R.wrap(greet, function(gr, name) {
>   return gr(name).toUpperCase();
> });
> shoutedGreet("Kathy"); //=> "HELLO KATHY"
>
> var shortenedGreet = R.wrap(greet, function(gr, name) {
>   return gr(name.substring(0, 3));
> });
> shortenedGreet("Robert"); //=> "Hello Rob"
> ```

Both examples are better expressed *without* [`R.wrap`][1]:

```javascript
var shoutedGreet = R.compose(R.toUpper, greet);
var shortenedGreet = R.compose(greet, R.take(3));
```

If we can't come up with a compelling example we should consider deprecating the function.

[1]: http://ramdajs.com/docs/#wrap

**Comment by asaf-romano** [comment]
For transforming the first argument of a function compose/pipe is better than both wrap and useWith. But for modifying the second argument, your choices are useWith and wrap, and it's hard to choose between them. With useWith you get much shorter, point-free code, but with wrap you get something that's often more expressive and clear at first glance, especially for someone who doesn't know the library very well.

**Comment by CrossEye** [comment]
The trouble with trying to come up with better examples for this is that to my mind, it's there chiefly to help introduce side-effects, not things I want to demonstrate in our documentation.  Here's a poor-man's version of @raine's [treis][tr]:

```js
var log = msg => fn => R.wrap(fn, function(f, ...args) {
  var result = f.apply(this, args);
  console.log({
    msg,
    input: R.map(R.toString, args),
    output: result
  });
  return result;
});

log("adding")(R.add)(19, 23); //=> 42, but logs {msg: 'adding', input: [19, 23], output: 42}
```

I could also see it being used for arbitrary decorators:

```js
var nullGuard = (msg, positions) => fn => R.wrap(fn, function(f, ...args) {
  if (R.anyPass(R.map(R.pipe(R.nth, R.isNil), positions), args)) {
      throw new Error(msg);
  }
  return f.apply(this, args);
});
```

It's probably best used for cases like that first one, where you want to deal with the inputs, the function, and the output, possibly even by deciding _not_ to call the function.

But I can't think of simple documentation-ready examples for this.  And I agree that this is a problem.  We really should be able to document our functions reasonably.

But I think this is a useful function in the same way `tap` is useful.  


  [tr]: https://github.com/raine/treis

**Comment by davidchambers** [comment]
> ```javascript
> var log = msg => fn => R.wrap(fn, function(f, ...args) {
>   var result = f.apply(this, args);
>   console.log({
>     msg,
>     input: R.map(R.toString, args),
>     output: result
>   });
>   return result;
> });
>
> log("adding")(R.add)(19, 23); //=> 42, but logs {msg: 'adding', input: [19, 23], output: 42}
> ```

Handing a function to `R.wrap` only for it to hand it back to us is a roundabout way of setting a function's arity. This can be done explicitly via `R.curryN`:

```javascript
var log = msg => fn => R.curryN(fn.length, () => {
  var result = fn.apply(this, arguments);
  console.log({
    msg: msg,
    input: R.map(R.toString, arguments),
    output: result
  });
  return result;
});
```

If the `length` of the returned function is not important, we don't need `R.curryN`:

```javascript
var log = msg => fn => () => {
  var result = fn.apply(this, arguments);
  console.log({
    msg: msg,
    input: R.map(R.toString, arguments),
    output: result
  });
  return result;
};
```

Writing decorators in vanilla JavaScript is quite natural. I don't see the appeal of `R.wrap`, but I may be overlooking a significant factor.

**Comment by CrossEye** [comment]
You may be right.  It's certainly not essential.

And that example is something that came to my head because I'd used it that way before.  I tried to simplify my code by making the log message external (rather than use something a bit more like [treis][tr] with the concomitant added complexity of finding function names or synthesizing a default one.)  That may have confused the issue.

The API might look like this:

```js
var someFunc = (...params) => {/* ... */ return someVal;}
someFunc = R.wrap(someFunc, logger);

// ...

var someModule = {
  // ...
}
// someModule = <iterate over functions in someModule, 
//                           replacing values with R.wrap(value, logger)>
``` 

Basically, I think of it as an easily way not simply to write function decorators but also to make it clear what you are doing.  To me this is entirely about (perhaps some analog to) Aspect-Oriented programming.

  [tr]: https://github.com/raine/treis

**Comment by davidchambers** [comment]
This issue is no longer relevant now that #1818 has been merged.
